CASSJAVA-52: Do not always cleanup Guava shaded module before packaging

lukasz-antoniak opened the pull request:
Fixes [CASSJAVA-52](https://issues.apache.org/jira/browse/CASSJAVA-52).

ASF CI performs the following sequence of operations to compile the driver and run integration tests:
```
$ mvn clean install -DskipTests
$ java11 (or any other version that we wish to run the tests with)
$ mvn verify -DccmJvm.version=11 -Dccm.version=4.0.4 -DtestJavaHome=...
```

`mvn verify` should not need to recompile source code, because JVM version is not set to 1.8 any more. Build failure:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project java-driver-guava-shaded: Compilation failure
[ERROR] /cassandra-java-driver/guava-shaded/src/main/java/com/google/common/primitives/LexicographicalComparatorHolderSubstitution.java:[24,17] cannot access java.util.Comparator
[ERROR]   bad class file: /modules/java.base/java/util/Comparator.class
[ERROR]     class file has wrong version 55.0, should be 53.0
[ERROR]     Please remove or make sure it appears in the correct subdirectory of the classpath.
```
CI failed job: [link](https://ci-cassandra.apache.org/blue/organizations/jenkins/cassandra-java-driver/detail/4.x/21/pipeline/).

SiyaoIsHiding reviewed (approved):
Confirm that it fixes the build failure.

absurdfarce commented:
Something about this seems... wrong to me.  I can't quite put my finger on it yet, but this feels like a fix for this specific situation rather than something we'd want to do generally.

Wouldn't I normally expect that "mvn verify" or "mvn package" (or anything that has to actually do a package op generally) would clean out whatever classes might happen to be around and start fresh?

It feels like what we actually want here is some way to tell "mvn verify" that hey, it's okay to use whatever class files you find... you don't need to start from scratch.  Maybe make the clean operation conditional on some Maven property and include that prop in the build?

Furthermore... don't we wind up with a Guava shaded artifact in the local Maven cache after "mvn install" completes?  If that's the case why is "mvn verify" trying to compile _anything_... why isn't it just using what's in the cache?

Something still seems wrong here....

lukasz-antoniak commented:
> Wouldn't I normally expect that "mvn verify" or "mvn package" (or anything that has to actually do a package op generally) would clean out whatever classes might happen to be around and start fresh?

Maven will check if any of source files has changed, and if not will not attempt to recompile. That is a standard behaviour. To explicitly "start from beginning" you would need to add `clean` option. Driver requires a compilation with JDK 1.8, so this version is configured during `mvn clean install`. Then we switch JDK to newer one to run integration tests. Compiled classes are left untouched, generated by JDK 1.8.

> Furthermore... don't we wind up with a Guava shaded artifact in the local Maven cache after "mvn install" completes? If that's the case why is "mvn verify" trying to compile anything... why isn't it just using what's in the cache?

The JAR in local cache is present, but this module is a project module, so Maven needs to try to compile it. The whole issue here is that we always asked it to clean compiled classes and recompile sources again. If we apply the PR and remove the custom fragment of POM, Maven will not attempt to recompile Guava shaded module once compiled sources are up to date.

Note that no other module uses `maven-clean-plugin` to clean target directory before package phase.

absurdfarce commented:
@lukasz-antoniak and I talked through this pretty extensively this afternoon.  After digging in a bit it looks like the underlying issue here is that the Maven shade plugin moves things around in it's target directory in order to create the structure you want for the shaded JAR.  The guava-shaded project also contains two Java source files that need to be included in the shaded JAR, and it's the intersection of these two behaviours that leads to the problem:

* "mvn clean install" compiles everything, including the shaded JAR, and installs it into the local Maven repo
* We switch to a new Java version (Java11 in this case)
* We then come along and do a "mvn verify", but when Maven checks to see whether we have built class files for the two Java classes that are part of that project it can't find them (because the shade plugin in the previous step moved them around) so it tries to recompile them.  We thus find JVM classes with the wrong class version.

We're pretty confident in this diagnosis; we literally watched these directories get moved around in real time on the file system while the build was happening.

Based on testing @lukasz-antoniak indicates that the fix offered in this PR (removing the clean step from the shaded build on a package) will resolve the issue.  It's not at all clear how _excluding_ this clean would fix anything, though, and I'm reluctant to make a build fix which relies on a mechanism we don't understand.

Another option we discussed was seeing if we could move the source files in the lib somewhere else; that would leave guava-shaded as supporting only shaded artifacts, freeing the shade plugin to move things around at will.  The two class files in question are Graal substitutions that came in with [JAVA-2799](https://datastax-oss.atlassian.net/browse/JAVA-2799).  In addition to this ticket there's a lot of backstory there to consider:

* [JAVA-2782](https://datastax-oss.atlassian.net/browse/JAVA-2782) which documents the original source of the problem
* The cassandra-quarkus [issue](https://github.com/datastax/cassandra-quarkus/issues/25) which JAVA-2782 was migrated to
* The [original attempt](https://github.com/apache/cassandra-java-driver/pull/1443) to fix this issue directly in the Java driver
* The [subsequent attempt](https://github.com/datastax/java-driver-shaded-guava/pull/4) which actually did fix this in the old java-driver-guava-shaded lib

If we want to move this functionality somewhere else we should make sure we aren't breaking any functionality discussed in that history.

lukasz-antoniak commented:
I have diagnosed why removing Maven clean plugin solves the problem.

When I run `mvn verify` phase and `maven-clean-plugin` is not present, we can notice the following log:
```
[DEBUG] New compile source roots:
  /Users/Developer/GitHub/cassandra-java-driver/guava-shaded/src/main/java
  /Users/Developer/GitHub/cassandra-java-driver/guava-shaded/target/generated-sources/annotations
[DEBUG] CompilerReuseStrategy: reuseCreated
[DEBUG] useIncrementalCompilation disabled
[INFO] Nothing to compile - all classes are up to date
```
On the other hand, if we run the same with `maven-clean-plugin` enabled, we see:
```
[DEBUG] New compile source roots:
  /Users/Developer/GitHub/cassandra-java-driver/guava-shaded/src/main/java
  /Users/Developer/GitHub/cassandra-java-driver/guava-shaded/target/generated-sources/annotations
[DEBUG] CompilerReuseStrategy: reuseCreated
[DEBUG] useIncrementalCompilation disabled
```
Log `Nothing to compile - all classes are up to date` indicates that all sources are compiled and up to date. This corresponds to this [part of code](https://github.com/apache/maven-compiler-plugin/blob/maven-compiler-plugin-3.8.1/src/main/java/org/apache/maven/plugin/compiler/AbstractCompilerMojo.java#L850). For a source file to be considered up-to-date compiled, it needs to ([source code](https://github.com/sonatype/plexus-compiler/blob/master/plexus-compiler-api/src/main/java/org/codehaus/plexus/compiler/util/scan/StaleSourceScanner.java#L94)):
1. Have a present corresponding target file.
2. Modification timestamp of compiled class later than source file.

Above logic has drawn my attention to the target classes and how clean plugin impacts the behaviour.

Maven clean plugin disabled:
```
$ java8
$ mvn clean install -DskipTests
$ ls -l ./guava-shaded/target/classes/com/google/common/primitives/LexicographicalComparatorHolderSubstitution.class
-rw-r--r--  1 lukasz staff  1498 Jan 10 09:26 ./guava-shaded/target/classes/com/google/common/primitives/LexicographicalComparatorHolderSubstitution.class
$ java11
$ mvn verify ....
$ ls -l ./guava-shaded/target/classes/com/google/common/primitives/LexicographicalComparatorHolderSubstitution.class
-rw-r--r--  1 lukasz staff  1498 Jan 10 09:26 ./guava-shaded/target/classes/com/google/common/primitives/LexicographicalComparatorHolderSubstitution.class
```

Maven clean plugin enabled:
```
$ java8
$ mvn clean install -DskipTests
$ ls -l ./guava-shaded/target/classes/com/google/common/primitives/LexicographicalComparatorHolderSubstitution.class
ls: ./guava-shaded/target/classes/com/google/common/primitives/LexicographicalComparatorHolderSubstitution.class: No such file or directory
$ ls -l ./guava-shaded/target/classes/com/                                                                          
total 0
drwxr-xr-x  3 lukasz staff  96 Jan 10 09:34 datastax
```

This shows that Maven clean plugin removes compiled `LexicographicalComparatorHolderSubstitution` class from `com/google/common/primitives` directory. This is **desired** as the class has been copied by shaded plugin to `com/datastax/oss/driver/shaded/guava/common/primitives/LexicographicalComparatorHolderSubstitution.class`. It becomes also clear why removing the clean plugin works - compiled source files under `target/com/google/common/primitives` are still present when running `mvn verify`.

If we remove Maven clean plugin, resulting shaded Guava JAR has two `LexicographicalComparatorHolderSubstitution` classes, whereas only one should be present.

lukasz-antoniak committed:
Conditionally compile shaded Guava module
patch by Lukasz Antoniak; reviewed by Alexandre Dutra and Bret McGuire for CASSJAVA-52

lukasz-antoniak commented:
@absurdfarce, I have been long trying to come up with a workaround. I see two options:
1. Add `-pl '!guava-shaded'` to `mvn verify` command not to compile shaded Guava module when running integration tests. It is one more parameter in the command line and may be disruptive to remember for other contributors.
2. Attempt to compile Guava shaded module only if we are running JVM 1.8 (added to this PR). This does not require any additional changes to CI scripts.

Thoughts?

absurdfarce commented:
Hmmm, okay, I didn't quite have my understanding of this right.

My working assumption was that the shade plugin would _move_ the compiled class files to the new location rather than copy them.  If you don't do this you potentially have duplication of class files if your build specified the original target directory (which would be common) as well as a new class directory created and added by the shade plugin.  Under that scenario you wind up with missing class files either way; either the source files are missing from the original target directory because they were moved (by the shade plugin) or because they were deleted (by the clean plugin).

If the shade plugin is only _copying_ files then yes, I agree with your analysis.  Removing source files from the original target directory would trigger recompilation... which is exactly what we're seeing.

absurdfarce commented:
After pondering this a bit more I think I'm okay with this as a solution to the immediate problem.  At the end of the day we'll have a problem if we don't keep the Java8 compiled class files around.  If we want to test with other JVMs we'll need to be able to switch to those JVMs and we can't have recompilation in that case.

Actually... wait... why is that true?  If we switch to a Java11 JDK why aren't we building Java8-compatible class files anyway?  We... specify that we want the [target version produced by the compiler](https://github.com/apache/cassandra-java-driver/blob/4.x/pom.xml#L608) to be 1.8... so how are we winding up with version 55 code _at all_?

I'm still gonna approve this because I agree we have a much clearer understanding of the mechanism now but it still feels like something is missing here.

absurdfarce commented:
Bah, that's what I get for rushing... I don't think I'd looked at your most recent code changes @lukasz-antoniak.  I like those much better on their face anyway.  I'm still really unclear as to how we're ever winding up with version 55 class files but there's no question your most recent proposal is much more elegant.

absurdfarce commented:
Kicking off DataStax CI run to make sure we don't have any regressions here.  I'm expecting it'll be Just Fine ™️

absurdfarce commented:
DataStax CI run was clean, a couple of completely unrelated failures for tests known to fail intermittently.  Calling this good.

absurdfarce commented:
@lukasz-antoniak can you rebase these to a single commit and add the appropriate "patch by" line?

lukasz-antoniak commented:
Done.
